pxr/imaging/hdx/shaders/renderPass.glslfx

```diff
@@ -192,11 +261,70 @@ void RenderOutput(vec4 Peye, vec3 Neye, vec4 color, vec4 patchCoord)
     int instanceId = GetDrawingCoord().instanceIndex[0];
     instanceIdOut = IntToVec4(instanceId);
 
-    elementIdOut = IntToVec4(GetElementID());
-    edgeIdOut = IntToVec4(GetPrimitiveEdgeId());
-    pointIdOut = IntToVec4(GetPointId());
+    int elementId = GetElementID();
+    int edgeId = GetPrimitiveEdgeId();
+    int pointId = GetPointId();
+
+    elementIdOut = IntToVec4(elementId);
+    edgeIdOut = IntToVec4(edgeId);
+    pointIdOut = IntToVec4(pointId);
     
     neyeOut = IntToVec4(hd_vec4_2_10_10_10_set(vec4(Neye,0)));
+
+    // add the item to the pick buffer, but only unique
+#if defined(HD_HAS_PickBuffer)
+
+    // if the pick buffer is not initialized, we are done 
+    if (PickBuffer[0] == 0)
+        return;
+	
+    // prepare some constants
+    const int entrySize = 3;
+    const int numSubBuffers = PickBuffer[0];
+    const int subBufferCapacity = PickBuffer[1];
+    const int tableOffset = PickBuffer[2];
+    const int storageOffset = PickBuffer[3];
+    const int partId =  PickBuffer[4] * elementId + 
+                        PickBuffer[5] * edgeId +
+                        PickBuffer[6] * pointId;
+
+    // calculate the item's hash
+    const int fnv1Seed = -2128831035;
+    int fnv1Value = fnv1Hash(primId, fnv1Seed);
+    fnv1Value = fnv1Hash(instanceId, fnv1Value);
+    fnv1Value = fnv1Hash(partId, fnv1Value);
+
+    // more constants
+    const int subBufferNumber = fnv1Value % numSubBuffers;
+    const int sizeOffset = tableOffset + subBufferNumber;
+    const int subBufferOffset = storageOffset + 
+                subBufferNumber * subBufferCapacity * entrySize;
+
+    // loop through the item's sub-buffer to see if the item is already there
+    // if not, add it in atomic fashion
+    int entryNumber = 0;
+    bool keepMoving = true;
+    do
+    {
+        // see if we need to add a new entry
+        if (entryNumber == PickBuffer[sizeOffset]) {
+            atomicCompSwap(PickBuffer[sizeOffset], entryNumber, entryNumber + 1);
+        }
+
+        // if either the item equals to the current entry or we managed to 
+        // initialize the entry with the item's data, we are done
+        if (compareOrSet(
+                subBufferOffset + entryNumber * entrySize, 
+                primId, instanceId, partId)) {
+            keepMoving = false;
+        }
+        else if(++entryNumber == subBufferCapacity) {
```

Comment:
Not important, but if you move this into the while() condition, and make the "keepMoving" above a "break", it makes the loop code a bit smaller.